src/components/repeat/repeat.component.tsx

```diff
@@ -69,101 +67,109 @@ const Repeat: React.FC<FormFieldProps> = ({ question, onChange }) => {
-      const index = allFormFields.findIndex((f) => f.id === field);
-      allFormFields.splice(index, 1);
-      delete values[field];
-    });
+    setRows(rows.filter((q) => q.id !== question.id));
   };
 
-  const nodes = obsGroups.map((question, index) => {
-    const keyId = question.id + '-' + index;
-    const deleteControl =
-      obsGroups.length > 1 && encounterContext.sessionMode !== 'view' ? (
-        <div key={keyId} className={styles.removeButton}>
-          <Button
-            className={styles.button}
-            renderIcon={() => <TrashCan size={16} />}
-            kind="danger--tertiary"
-            onClick={() => removeNthRow(question)}>
-            <span>{t('removeGroup', 'Remove group')}</span>
-          </Button>
-        </div>
-      ) : null;
-
+  const nodes = useMemo(() => {
     return (
-      <div key={keyId}>
-        {index !== 0 && (
-          <div>
-            <hr className={styles.divider} />
-          </div>
-        )}
-        <div className={styles.obsGroupContainer}>
-          <ObsGroup
-            question={question}
-            onChange={onChange}
-            handler={formFieldHandlers[question.type]}
-            deleteControl={index !== 0 ? deleteControl : null}
-          />
-        </div>
-      </div>
-    );
-  });
-
-  encounterContext.sessionMode != 'view' &&
-    nodes.push(
-      <div key={id}>
-        {showAddButton(question.questionOptions.repeatOptions?.limit, counter) && (
-          <Button
-            className={styles.button}
-            iconDescription={t('add', 'Add')}
-            renderIcon={() => <Add size={16} />}
-            kind="tertiary"
-            onClick={() => {
-              const nextCount = counter + 1;
-              handleAdd(counter);
-              setCounter(nextCount);
-            }}>
-            <span>{question.questionOptions.repeatOptions?.addText || `${t('add', 'Add')}`}</span>
-          </Button>
-        )}
-      </div>,
+      fieldComponent &&
+      rows.map((question, index) => {
+        const component = <fieldComponent.Component question={question} onChange={onChange} handler={handler} />;
+        return (
+          component && (
+            <div key={question.id + '_wrapper'}>
+              {index !== 0 && (
+                <div>
+                  <hr className={styles.divider} />
+                </div>
+              )}
+              {isGrouped ? <div className={styles.obsGroupContainer}>{component}</div> : component}
+              <RepeatControls
+                question={question}
+                rows={rows}
+                questionIndex={index}
+                handleDelete={() => {
+                  removeNthRow(question);
+                }}
+                handleAdd={() => {
+                  const nextCount = counter + 1;
+                  handleAdd(nextCount);
+                  setCounter(nextCount);
+                }}
+              />
+            </div>
+          )
```

Comment:
nit: Do we need to render `null` here as a fallback if `component` doesn't[ resolve to a boolean value](https://kentcdodds.com/blog/use-ternaries-rather-than-and-and-in-jsx)?

```ts
component ? (
  // ... code
) : null)
```